File: rust-mcp/src/capture.rs
use std::{
    path::PathBuf,
    sync::{
        Arc,
        atomic::{AtomicU64, Ordering},
    },
    time::{Duration, Instant, SystemTime, UNIX_EPOCH},
};

use anyhow::{Context, Result, bail};
use serde_json::{Value, json};
use sha2::{Digest, Sha256};
use tokio::{fs, sync::Semaphore};
use tokio_util::sync::CancellationToken;

use crate::{
    Config,
    process::{ProcessError, ProcessOptions, spawn_process},
};

const RETRY_BACKOFF: Duration = Duration::from_millis(150);
static UNIQUE_COUNTER: AtomicU64 = AtomicU64::new(0);

#[derive(Debug, Clone)]
pub struct CaptureResult {
    pub image: Vec<u8>,
    pub mime_type: String,
    pub metadata: Value,
}

#[derive(Debug, Clone)]
pub struct CaptureService {
    config: Arc<Config>,
    gate: Arc<Semaphore>,
}

impl CaptureService {
    #[must_use]
    pub fn new(config: Arc<Config>) -> Self {
        Self {
            config,
            gate: Arc::new(Semaphore::new(1)),
        }
    }

    /// Capture the host virtual display using the bounded native capture worker.
    ///
    /// # Errors
    /// Returns validation, queue timeout, cancellation, worker, image-validation, or I/O errors.
    pub async fn display(
        &self,
        quality: u8,
        cancellation: CancellationToken,
    ) -> Result<CaptureResult> {
        validate_quality(quality)?;
        self.capture_with_policy(
            "display",
            vec!["display".to_owned(), quality.to_string()],
            cancellation,
        )
        .await
    }

    /// Capture the largest visible top-level window for a process/process tree.
    ///
    /// # Errors
    /// Returns validation, queue timeout, cancellation, worker, image-validation, or I/O errors.
    pub async fn program(
        &self,
        pid: u32,
        quality: u8,
        include_process_tree: bool,
        cancellation: CancellationToken,
    ) -> Result<CaptureResult> {
        if pid == 0 {
            bail!("pid must be a positive process ID.");
        }
        validate_quality(quality)?;
        self.capture_with_policy(
            "program",
            vec![
                "program".to_owned(),
                quality.to_string(),
                pid.to_string(),
                include_process_tree.to_string(),
            ],
            cancellation,
        )
        .await
    }

    async fn capture_with_policy(
        &self,
        operation: &str,
        worker_args: Vec<String>,
        cancellation: CancellationToken,
    ) -> Result<CaptureResult> {
        let queue_started = Instant::now();
        let queue_timeout =
            Duration::from_millis(self.config.screen_capture_queue_timeout_ms.max(1));
        let permit = tokio::select! {
            biased;
            () = cancellation.cancelled() => bail!("Screen capture cancelled while waiting for the capture worker."),
            result = tokio::time::timeout(queue_timeout, self.gate.acquire()) => {
                match result {
                    Ok(Ok(permit)) => permit,
                    Ok(Err(_)) => bail!("Screen capture queue was closed."),
                    Err(_) => bail!("Screen capture queue remained busy for {} ms. Retry shortly.", queue_timeout.as_millis()),
                }
            }
        };
        let queue_wait_ms = duration_ms(queue_started.elapsed());
        let attempts = self.config.screen_capture_retries.saturating_add(1).max(1);
        let attempt_timeout =
            Duration::from_millis(self.config.screen_capture_attempt_timeout_ms.max(1));
        let overall_timeout_ms =
            duration_ms(
                attempt_timeout
                    .saturating_mul(u32::try_from(attempts).unwrap_or(u32::MAX))
                    .saturating_add(RETRY_BACKOFF.saturating_mul(
                        u32::try_from(attempts.saturating_sub(1)).unwrap_or(u32::MAX),
                    )),
            );
        let mut last_error = None;
        let mut completed_attempt = 0_usize;

        for attempt in 1..=attempts {
            completed_attempt = attempt;
            match self
                .worker_attempt(&worker_args, attempt_timeout, cancellation.child_token())
                .await
            {
                Ok(mut capture) => {
                    let metadata = capture.metadata.as_object_mut().with_context(|| {
                        format!("{operation} capture worker metadata must be a JSON object")
                    })?;
                    metadata.insert("capture_attempts".to_owned(), json!(attempt));
                    metadata.insert("capture_retried".to_owned(), json!(attempt > 1));
                    metadata.insert("capture_queue_wait_ms".to_owned(), json!(queue_wait_ms));
                    metadata.insert(
                        "capture_attempt_timeout_ms".to_owned(),
                        json!(duration_ms(attempt_timeout)),
                    );
                    metadata.insert(
                        "capture_overall_timeout_ms".to_owned(),
                        json!(overall_timeout_ms),
                    );
                    drop(permit);
                    return Ok(capture);
                }
                Err(error) => {
                    let transient = is_transient_capture_error(&error);
                    last_error = Some(error);
                    if attempt >= attempts || !transient {
                        break;
                    }
                    tokio::select! {
                        () = cancellation.cancelled() => bail!("Screen capture cancelled before retry."),
                        () = tokio::time::sleep(RETRY_BACKOFF) => {},
                    }
                }
            }
        }
        drop(permit);
        let mut error = last_error.unwrap_or_else(|| anyhow::anyhow!("Screen capture failed."));
        error = error.context(format!(
            "Screen capture failed after {completed_attempt} attempt(s); queue_wait_ms={queue_wait_ms}; attempt_timeout_ms={}; overall_timeout_ms={overall_timeout_ms}",
            duration_ms(attempt_timeout)
        ));
        Err(error)
    }

    async fn worker_attempt(
        &self,
        worker_args: &[String],
        timeout: Duration,
        cancellation: CancellationToken,
    ) -> Result<CaptureResult> {
        let executable =
            std::env::current_exe().context("resolve Rust MCP executable for capture worker")?;
        let temp_dir = temporary_capture_dir();
        fs::create_dir_all(&temp_dir).await?;
        let image_path = temp_dir.join("capture.bin");
        let mut args = vec![
            "--capture-worker".to_owned(),
            image_path.to_string_lossy().into_owned(),
        ];
        args.extend(worker_args.iter().cloned());
        let result = spawn_process(
            &executable.to_string_lossy(),
            &args,
            ProcessOptions {
                timeout: Some(timeout),
                max_capture_chars: Some(200_000),
                ..ProcessOptions::default()
            },
            cancellation,
        )
        .await;
        let outcome = async {
            let process = result.map_err(anyhow::Error::new)?;
            let mut metadata: Value = serde_json::from_str(process.stdout.trim())
                .context("parse Rust capture-worker metadata")?;
            let image = fs::read(&image_path)
                .await
                .with_context(|| format!("read capture image {}", image_path.display()))?;
            let mime_type = metadata["mime_type"]
                .as_str()
                .map(str::to_owned)
                .context("capture-worker metadata omitted mime_type")?;
            validate_image(&image, &mime_type)?;
            let object = metadata
                .as_object_mut()
                .context("capture-worker metadata must be a JSON object")?;
            object.insert("bytes".to_owned(), json!(image.len()));
            object.insert("sha256".to_owned(), Value::String(sha256_hex(&image)));
            Ok(CaptureResult {
                image,
                mime_type,
                metadata,
            })
        }
        .await;
        fs::remove_dir_all(&temp_dir).await.ok();
        outcome
    }
}

/// Hidden worker entry point used by the long-lived MCP to make capture attempts killable.
///
/// Arguments are `<output-path> display <quality>` or
/// `<output-path> program <quality> <pid> <include-process-tree>`.
///
/// # Errors
/// Returns argument, platform, capture, image-write, or metadata serialization errors.
pub async fn run_capture_worker(arguments: &[String]) -> Result<()> {
    let output_path = arguments
        .first()
        .map(PathBuf::from)
        .context("capture worker requires an output path")?;
    let mode = arguments.get(1).map(String::as_str).unwrap_or_default();
    let quality = arguments
        .get(2)
        .context("capture worker requires quality")?
        .parse::<u8>()
        .context("parse capture quality")?;
    validate_quality(quality)?;

    #[cfg(windows)]
    let capture = match mode {
        "display" => crate::windows_capture::capture_display(quality)?,
        "program" => {
            let pid = arguments
                .get(3)
                .context("program capture requires pid")?
                .parse::<u32>()
                .context("parse capture pid")?;
            let include_process_tree = arguments
                .get(4)
                .context("program capture requires include-process-tree")?
                .parse::<bool>()
                .context("parse include-process-tree")?;
            crate::windows_capture::capture_program(pid, quality, include_process_tree)?
        }
        other => bail!("unsupported capture-worker mode {other:?}"),
    };

    #[cfg(not(windows))]
    let capture = match mode {
        "display" => crate::posix_capture::capture_display(quality).await?,
        "program" => {
            let pid = arguments
                .get(3)
                .context("program capture requires pid")?
                .parse::<u32>()
                .context("parse capture pid")?;
            let include_process_tree = arguments
                .get(4)
                .context("program capture requires include-process-tree")?
                .parse::<bool>()
                .context("parse include-process-tree")?;
            crate::posix_capture::capture_program(pid, quality, include_process_tree).await?
        }
        other => bail!("unsupported capture-worker mode {other:?}"),
    };

    if let Some(parent) = output_path.parent() {
        fs::create_dir_all(parent).await?;
    }
    fs::write(&output_path, &capture.image)
        .await
        .with_context(|| format!("write capture image {}", output_path.display()))?;
    let mut metadata = capture.metadata;
    let object = metadata
        .as_object_mut()
        .context("native capture metadata must be an object")?;
    object.insert(
        "mime_type".to_owned(),
        Value::String(capture.mime_type.to_owned()),
    );
    println!("{}", serde_json::to_string(&metadata)?);
    Ok(())
}

fn validate_quality(quality: u8) -> Result<()> {
    if !(1..=100).contains(&quality) {
        bail!("quality must be between 1 and 100.");
    }
    Ok(())
}

fn validate_image(image: &[u8], mime_type: &str) -> Result<()> {
    match mime_type {
        "image/jpeg" => {
            if image.len() < 5
                || !image.starts_with(&[0xff, 0xd8, 0xff])
                || !image.ends_with(&[0xff, 0xd9])
            {
                bail!("Screen capture did not return a valid JPEG image.");
            }
        }
        "image/png" => {
            const PNG_SIGNATURE: &[u8] = &[0x89, b'P', b'N', b'G', 0x0d, 0x0a, 0x1a, 0x0a];
            if !image.starts_with(PNG_SIGNATURE) {
                bail!("Screen capture did not return a valid PNG image.");
            }
        }
        other => bail!("Screen capture returned unsupported MIME type {other:?}."),
    }
    Ok(())
}

fn is_transient_capture_error(error: &anyhow::Error) -> bool {
    if let Some(process) = error.downcast_ref::<ProcessError>()
        && process.timed_out
    {
        return true;
    }
    let message = error.to_string().to_ascii_lowercase();
    [
        "timed out",
        "timeout",
        "printwindow",
        "bitblt",
        "getdc",
        "desktop compositor",
        "capture did not return",
        "capture-worker metadata",
    ]
    .iter()
    .any(|needle| message.contains(needle))
}

fn temporary_capture_dir() -> PathBuf {
    let millis = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_or(0, |duration| duration.as_millis());
    let counter = UNIQUE_COUNTER.fetch_add(1, Ordering::Relaxed);
    std::env::temp_dir().join(format!(
        "devbox-rust-capture-{millis}-{}-{counter}",
        std::process::id()
    ))
}

fn sha256_hex(bytes: &[u8]) -> String {
    format!("{:x}", Sha256::digest(bytes))
}

fn duration_ms(duration: Duration) -> u64 {
    u64::try_from(duration.as_millis()).unwrap_or(u64::MAX)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn image_validation_rejects_wrong_magic() {
        assert!(validate_image(b"not jpeg", "image/jpeg").is_err());
        assert!(validate_image(b"not png", "image/png").is_err());
    }

    #[test]
    fn capture_retry_backoff_matches_javascript_policy() {
        assert_eq!(RETRY_BACKOFF, Duration::from_millis(150));
    }

    #[test]
    fn transient_classifier_distinguishes_window_discovery_from_capture_stalls() {
        assert!(is_transient_capture_error(&anyhow::anyhow!(
            "BitBlt failed"
        )));
        assert!(is_transient_capture_error(&anyhow::anyhow!(
            "capture timed out"
        )));
        assert!(!is_transient_capture_error(&anyhow::anyhow!(
            "process has no visible window"
        )));
    }
}

File: cpp-mcp/src/capture.cpp
#include "devbox/scoped_thread.hpp"
#include "devbox/capture.hpp"
#include "devbox/native.hpp"
#include <algorithm>
#include <charconv>
#include <csignal>
#include <iostream>
#ifndef _WIN32
#include <sys/stat.h>
#endif
namespace devbox {
namespace {
volatile std::sig_atomic_t capture_interrupted = 0;
void capture_signal(int) {
    capture_interrupted = 1;
}
void validate_quality(unsigned quality) {
    if (quality < 1 || quality > 100)
        throw Error("quality must be between 1 and 100.");
}
std::uint32_t parse_number(std::string_view text, const char* name) {
    std::uint32_t number = 0;
    const auto result = std::from_chars(text.data(), text.data() + text.size(), number);
    if (result.ec != std::errc() || result.ptr != text.data() + text.size())
        throw Error(std::string("parse capture ") + name);
    return number;
}
} // namespace
void validate_capture_image(std::span<const std::uint8_t> image, std::string_view mime) {
    if (mime == "image/jpeg") {
        if (image.size() < 5 || image[0] != 0xff || image[1] != 0xd8 || image[2] != 0xff ||
            image[image.size() - 2] != 0xff || image.back() != 0xd9)
            throw Error("Screen capture did not return a valid JPEG image.");
    } else if (mime == "image/png") {
        constexpr std::uint8_t signature[]{0x89, 'P', 'N', 'G', 13, 10, 26, 10};
        if (image.size() < sizeof(signature) ||
            !std::equal(std::begin(signature), std::end(signature), image.begin()))
            throw Error("Screen capture did not return a valid PNG image.");
    } else
        throw Error("Screen capture returned unsupported MIME type \"" + std::string(mime) + "\".");
}
bool transient_capture_error(const std::exception& error) {
    if (const auto* process = dynamic_cast<const ProcessError*>(&error); process && process->timed_out)
        return true;
    const auto message = lower(error.what());
    for (const auto* needle : {"timed out", "timeout", "printwindow", "bitblt", "getdc", "desktop compositor",
                               "capture did not return", "capture-worker metadata"})
        if (message.find(needle) != message.npos)
            return true;
    return false;
}
ImageCapture CaptureService::attempt(std::optional<std::uint32_t> pid, unsigned quality, bool tree,
                                     const Cancel& cancel) {
    const auto root = fs::temp_directory_path() / path_from_utf8("devbox-cpp-capture-" + uuid());
#ifdef _WIN32
    if (!fs::create_directory(root))
        throw Error("create private capture temporary directory");
#else
    if (::mkdir(root.c_str(), 0700) != 0)
        throw Error("create private capture temporary directory");
#endif
    ScopeExit cleanup([&] {
        std::error_code ec;
        fs::remove_all(root, ec);
    });
    const auto path = root / "capture.bin";
    std::vector<std::string> args{"--capture-worker", path_text(path), pid ? "program" : "display",
                                  std::to_string(quality)};
    if (pid) {
        args.push_back(std::to_string(*pid));
        args.push_back(tree ? "true" : "false");
    }
    ProcessOptions options;
    options.timeout = Millis(std::max<std::uint64_t>(1, config_->screen_capture_attempt_timeout_ms));
    options.max_capture_chars = 200000;
    auto result = spawn_process(path_text(executable_path()), args, options, cancel);
    Json metadata;
    try {
        metadata = Json::parse(trim(result.stdout_text));
    } catch (const std::exception&) {
        throw Error("parse C++ capture-worker metadata");
    }
    if (!metadata.is_object())
        throw Error("capture-worker metadata must be a JSON object");
    const auto mime = json_string(metadata, "mime_type");
    if (mime.empty())
        throw Error("capture-worker metadata omitted mime_type");
    const auto bytes = read_file(path, 64 * 1024 * 1024);
    std::vector<std::uint8_t> image(bytes.begin(), bytes.end());
    validate_capture_image(image, mime);
    metadata["bytes"] = image.size();
    metadata["sha256"] = sha256(image);
    return {std::move(image), mime, std::move(metadata)};
}
asio::awaitable<ImageCapture> CaptureService::capture(std::optional<std::uint32_t> pid, unsigned quality,
                                                      bool tree, Cancel cancel) {
    validate_quality(quality);
    if (pid && *pid == 0)
        throw Error("pid must be a positive process ID.");
    const auto start = Clock::now();
    const auto queue_timeout = Millis(std::max<std::uint64_t>(1, config_->screen_capture_queue_timeout_ms));
    for (;;) {
        if (cancel && cancel->cancelled())
            throw Error("Screen capture cancelled while waiting for the capture worker.");
        bool available = false;
        if (occupied_.compare_exchange_strong(available, true))
            break;
        if (Clock::now() - start >= queue_timeout)
            throw Error("Screen capture queue remained busy for " + std::to_string(queue_timeout.count()) +
                        " ms. Retry shortly.");
        co_await async_delay(Millis(10), cancel);
    }
    ScopeExit release([this] { occupied_.store(false); });
    const auto waited = std::chrono::duration_cast<Millis>(Clock::now() - start).count();
    const auto attempts = std::max<std::size_t>(1, config_->screen_capture_retries + 1);
    const auto timeout = std::max<std::uint64_t>(1, config_->screen_capture_attempt_timeout_ms);
    const auto overall = timeout * attempts + 150 * (attempts - 1);
    std::exception_ptr last;
    std::size_t completed = 0;
    for (std::size_t index = 1; index <= attempts; ++index) {
        completed = index;
        bool retry = false;
        try {
            auto pending = workers_.run(
                [this, pid, quality, tree, cancel] { return attempt(pid, quality, tree, cancel); }, cancel);
            auto value = co_await std::move(pending);
            value.metadata["capture_attempts"] = index;
            value.metadata["capture_retried"] = index > 1;
            value.metadata["capture_queue_wait_ms"] = waited;
            value.metadata["capture_attempt_timeout_ms"] = timeout;
            value.metadata["capture_overall_timeout_ms"] = overall;
            co_return value;
        } catch (const std::exception& e) {
            last = std::current_exception();
            retry = transient_capture_error(e);
        }
        if (!retry || index == attempts)
            break;
        if (cancel && cancel->cancelled())
            throw Error("Screen capture cancelled before retry.");
        co_await async_delay(Millis(150), cancel);
    }
    try {
        if (last)
            std::rethrow_exception(last);
    } catch (const ProcessError&) {
        throw;
    } catch (const std::exception&) {
    }
    throw Error("Screen capture failed after " + std::to_string(completed) + " attempt(s); queue_wait_ms=" +
                std::to_string(waited) + "; attempt_timeout_ms=" + std::to_string(timeout) +
                "; overall_timeout_ms=" + std::to_string(overall));
}
int run_capture_worker(const std::vector<std::string>& arguments) {
    try {
        if (arguments.size() != 3 && arguments.size() != 5)
            throw Error("capture worker requires output path, mode, quality, and optional pid/process-tree");
        const auto output = path_from_utf8(arguments[0]);
        const auto quality = parse_number(arguments[2], "quality");
        validate_quality(quality);
        std::optional<std::uint32_t> pid;
        bool tree = true;
        if (arguments[1] == "program" && arguments.size() == 5) {
            pid = parse_number(arguments[3], "pid");
            if (arguments[4] != "true" && arguments[4] != "false")
                throw Error("parse include-process-tree");
            tree = arguments[4] == "true";
        } else if (arguments[1] != "display" || arguments.size() != 3)
            throw Error("unsupported capture-worker mode");
        auto cancel = std::make_shared<Cancellation>();
#ifndef _WIN32
        capture_interrupted = 0;
        const auto old_int = std::signal(SIGINT, capture_signal);
        const auto old_term = std::signal(SIGTERM, capture_signal);
        ScopeExit restore([&] {
            std::signal(SIGINT, old_int);
            std::signal(SIGTERM, old_term);
        });
        ScopedThread signals([cancel](ThreadStopToken stop) {
            while (!stop.stop_requested()) {
                if (capture_interrupted) {
                    cancel->cancel();
                    return;
                }
                std::this_thread::sleep_for(Millis(10));
            }
        });
#endif
        auto value = native_capture(pid, quality, tree, cancel);
        cancel->check();
        validate_capture_image(value.image, value.mime_type);
        if (!output.parent_path().empty())
            fs::create_directories(output.parent_path());
        write_file(output,
                   std::string_view(reinterpret_cast<const char*>(value.image.data()), value.image.size()));
        value.metadata["mime_type"] = value.mime_type;
        std::cout << value.metadata.dump() << '\n';
        return 0;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << '\n';
        return 1;
    }
}
} // namespace devbox
